// This is the complete, correct, unabridged, and de-duplicated proto file.
syntax = "proto2";
package pblczero;

message EngineVersion {
  optional uint32 major = 1;
  optional uint32 minor = 2;
  optional uint32 patch = 3;
}

message Weights {
  message Layer { optional float min_val = 1; optional float max_val = 2; optional bytes params = 3; }
  message ConvBlock { optional Layer weights = 1; optional Layer biases = 2; optional Layer bn_means = 3; optional Layer bn_stddivs = 4; optional Layer bn_gammas = 5; optional Layer bn_betas = 6; }
  message RMSNormParameters { optional Layer gammas = 1; optional Layer betas = 2; }
  message SEunit { optional Layer w1 = 1; optional Layer b1 = 2; optional Layer w2 = 3; optional Layer b2 = 4; }
  message Residual { optional ConvBlock conv1 = 1; optional ConvBlock conv2 = 2; optional SEunit se = 3; }

  // --- Flexible Encoder Block Definitions ---
  message AttentionBlock {
      optional Layer q_w=1; optional Layer q_b=2; optional Layer k_w=3; optional Layer k_b=4;
      optional Layer v_w=5; optional Layer v_b=6; optional Layer dense_w=7; optional Layer dense_b=8;
      optional Layer dense1_w=9; optional Layer dense1_b=10; optional Layer dense2_w=11; optional Layer dense2_b=12;
      optional RMSNormParameters ln1=13; optional RMSNormParameters ln2=14;
  }
  
  message Mamba2Block {
      optional RMSNormParameters norm=1; optional Layer in_proj_weights=2; optional Layer conv1d_weights=3;
      optional Layer conv1d_biases=4; optional Layer x_proj_weights=5; optional Layer dt_proj_weights=6;
      optional Layer dt_proj_biases=7; optional Layer A_log=8; optional Layer D=9; optional Layer out_proj_weights=10;
  }

  // This 'oneof' is the key to supporting multiple architectures cleanly.
  message EncoderBlock { oneof block_type { AttentionBlock attention=1; Mamba2Block mamba2=2; } }

  // --- Head Definitions ---
  message PolicyHead { optional Layer ip_pol_w=1; optional Layer ip_pol_b=2; optional Layer ip2_pol_w=3; optional Layer ip2_pol_b=4; optional Layer ip3_pol_w=5; optional Layer ip3_pol_b=6; optional Layer ip4_pol_w=7; }
  message ValueHead { optional Layer ip_val_w=1; optional Layer ip_val_b=2; optional Layer ip1_val_w=3; optional Layer ip1_val_b=4; optional Layer ip2_val_w=5; optional Layer ip2_val_b=6; }
  message PolicyHeads { optional PolicyHead vanilla=3; }
  message ValueHeads { optional ValueHead winner=1; optional ValueHead q=2; optional ValueHead st=3; }

  // --- Main Weights Structure ---
  optional ConvBlock input = 1;
  repeated Residual residual = 2;
  repeated EncoderBlock encoder = 27; // Use the new flexible encoder
  optional uint32 headcount = 28;
  optional PolicyHeads policy_heads = 45;
  optional ValueHeads value_heads = 44;
  optional Layer ip_mov_w=31; optional Layer ip_mov_b=32; optional Layer ip1_mov_w=13; optional Layer ip1_mov_b=14; optional Layer ip2_mov_w=15; optional Layer ip2_mov_b=16;
}

message TrainingParams { optional uint32 training_steps = 1; optional float learning_rate = 2; optional float mse_loss = 3; optional float policy_loss = 4; optional float accuracy = 5; }

message NetworkFormat {
  enum InputFormat { INPUT_UNKNOWN=0; INPUT_CLASSICAL_112_PLANE=1; INPUT_112_WITH_CASTLING_PLANE=2; INPUT_112_WITH_CANONICALIZATION=3; INPUT_112_WITH_CANONICALIZATION_HECTOPLIES=4; INPUT_112_WITH_CANONICALIZATION_V2=5; }
  enum OutputFormat { OUTPUT_UNKNOWN=0; OUTPUT_CLASSICAL=1; OUTPUT_WDL=2; }
  enum NetworkStructure { NETWORK_UNKNOWN=0; NETWORK_CLASSICAL_WITH_HEADFORMAT=3; NETWORK_SE_WITH_HEADFORMAT=4; NETWORK_ATTENTIONBODY_WITH_MULTIHEADFORMAT=134; NETWORK_MAMBA2_WITH_HEADFORMAT=16; }
  enum PolicyFormat { POLICY_UNKNOWN=0; POLICY_CLASSICAL=1; POLICY_CONVOLUTION=2; POLICY_ATTENTION=3; }
  enum ValueFormat { VALUE_UNKNOWN=0; VALUE_CLASSICAL=1; VALUE_WDL=2; }
  enum MovesLeftFormat { MOVES_LEFT_NONE=0; MOVES_LEFT_V1=1; }
  enum ActivationFunction { DEFAULT_ACTIVATION_RELU=0; DEFAULT_ACTIVATION_MISH=1; }
  enum InputEmbeddingFormat { INPUT_EMBEDDING_NONE=0; INPUT_EMBEDDING_PE_MAP=1; INPUT_EMBEDDING_PE_DENSE=2; }

  optional InputFormat input=1;
  optional OutputFormat output=2;
  optional NetworkStructure network=3;
  optional PolicyFormat policy=4;
  optional ValueFormat value=5;
  optional MovesLeftFormat moves_left=6;
  optional ActivationFunction default_activation=7 [default=DEFAULT_ACTIVATION_RELU];
  optional InputEmbeddingFormat input_embedding=10;

  // New Configuration Fields for Mamba2 and RoPE
  optional bool rope_2d_enabled=11;
  optional float rope_2d_theta=12;
  optional int32 mamba2_d_state=14;
  optional int32 mamba2_d_conv=15;
  optional int32 mamba2_expand_factor=16;
  optional int32 mamba2_num_heads=17;
  optional int32 mamba2_head_dim=18;
  optional int32 mamba2_chunk_size=19;
  optional bool mamba2_use_norm=20;
  optional int32 mamba2_dt_rank=22;
}

message Format {
  enum Encoding { UNKNOWN=0; LINEAR16=1; }
  optional Encoding weights_encoding=1;
  optional NetworkFormat network_format=2;
}

message Net {
  optional fixed32 magic=1;
  optional EngineVersion min_version=3;
  optional Format format=4;
  optional TrainingParams training_params=5;
  optional Weights weights=10;
}
